runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,481 @@
+static inline struct defer_node *pop_bulk_defer_list(struct tls_node *node, uint64_t epoch) {
+  struct defer_node *dnode = node->deferList, *prev = NULL;
+  while (dnode != NULL) {
+    if (epoch >= dnode->targetEpoch) {
+      break;
+    }
+
+    prev = dnode;
+    dnode = dnode->next;
+  }
+
+  // If both nodes are NULL, then the list was actually empty
+  if (dnode == NULL && prev == NULL) {
+    return NULL;
+  }
+
+  // If prev is NULL then we have the actual head of the list
+  if (prev == NULL) {
+    node->deferList = NULL;
+    return dnode;
+  }
+
+  // If neither are NULL, we need to correct list so previous points to NULL
+  prev->next = NULL;
+  return dnode;
+}
+
+// Requires spinlock.
+static inline struct defer_node *pop_all_defer_list(struct tls_node *node);
+static inline struct defer_node *pop_all_defer_list(struct tls_node *node) {
+  struct defer_node *head = node->deferList;
+  node->deferList = NULL;
+  return head;
+}
+
+// Requires spinlock.
+static inline void push_defer_list(struct tls_node *node, struct defer_node *dnode);
+static inline void push_defer_list(struct tls_node *node, struct defer_node *dnode) {
+  dnode->next = node->deferList;
+  node->deferList = dnode;
+}
+
+// Requires spinlock.
+static inline struct defer_node *get_recycle_list(struct tls_node *node);
+static inline struct defer_node *get_recycle_list(struct tls_node *node) {
+  return node->recycleList;
+}
+
+// Requires spinlock.
+static inline struct defer_node *pop_recycle_list(struct tls_node *node);
+static inline struct defer_node *pop_recycle_list(struct tls_node *node) {
+  struct defer_node *dnode = node->recycleList;
+  
+  // If empty, return fresh allocation
+  if (dnode == NULL) {
+    return chpl_mem_calloc(1, sizeof(struct defer_node), 0, 0, 0);
+  }
+
+  node->recycleList = dnode->next;
+  return dnode;
+}
+
+static inline void push_recycle_list(struct tls_node *node, struct defer_node *dnode);
+static inline void push_recycle_list(struct tls_node *node, struct defer_node *dnode) {
+  dnode->next = node->recycleList;
+  node->recycleList = dnode;
+}
+
+static inline void delete_data(struct defer_node *dnode);
+static inline void delete_data(struct defer_node *dnode) {
+  if (dnode->numData) {
+      void **arr = dnode->data;
+      for (int i = 0; i < dnode->numData; i++) {
+          chpl_mem_free(arr[i], 0, 0);
+      }
+  }
+  chpl_mem_free(dnode->data, 0, 0);
+}
+
+// Initializes TLS; should only need to be called once.
+static void init_tls(void);
+static void init_tls(void) {
+  struct tls_node *node = chpl_calloc(1, sizeof(struct tls_node));
+  node->epoch = get_global_epoch();
+
+  // Append to head of list
+  struct tls_node *old_head;
+  do {
+    old_head = chpl_qsbr_tls_list;
+    node->next = old_head;
+  } while (!atomic_compare_exchange_weak_uintptr_t(
+      (uintptr_t *) &chpl_qsbr_tls_list, (uintptr_t) old_head, (uintptr_t) node)
+    );
+
+  CHPL_TLS_SET(chpl_qsbr_tls, node);
```

Comment:
Should threads that are no longer running remove themselves from this TLS list? I'm not even sure how to do that... but I wonder if some usage models would create many pthreads and then pollute this list with useless entries that are never reclaimed?

Reply:
Wait... oh, I didn't realize tasking layers actually discard threads, I thought they were all pooled? The way to handle that is similar to hazard pointers where we keep an active flag that is atomic and have new threads scan the list for any non-active nodes before creating a new one which allows recycling without memory reclamation. However, I assumed that it wouldn't be the case, but is it possible in any of the tasking layers?

Reply:
I actually don't know, but I think we need to at least document the assumptions.
Certainly one could use the foreign function interface to start/end pthreads directly...
I think this is worth asking @ronawho about. Elliot - the particular case here is a data structure with an element per thread, but the elements are never removed. Is that OK, and if not, where could we trigger marking the element for a particular thread as no longer needed / removing it? (e.g. could we call use pthread_key_create to arrange for a function to be called when the thread exits, but that assumes that the tasking layer is actually backed by pthreads or compatible with them).

Reply:
I can add a way for other threads to register themselves and just allow recycling. It doesn't matter what kind of threading abstraction they use, as long as they have some kind of TLS since it's needed to find their tls_node. 

For external pthreads we can just set it's nTasks to 1 for so long as it exists. If it never gets unregistered that's a programmer error. If they never actually need access to protected data, they don't need to register, same for any thread (only issue is that it's hard to reliably determine which thread  will or will not indirectly access protected data, but the user can).

Reply:
> If they never actually need access to protected data, they don't need to register, same for any thread (only issue is that it's hard to reliably determine which thread will or will not indirectly access protected data, but the user can).

As far as I know, we can reasonably assume that only Chapel tasks can participate in the Quiescent State-Based Reclamation.

So, that gives me an idea that might finesse the issue. In fact I think it already works this way. What if only threads that actually run tasks are registered? As far as I know, any thread that runs tasks will be something the implementation wants to pool and limit the number of. As it is, init_tls is only called in the functions we're arranging to be called by the Chapel tasking layers. So an "external" pthread wouldn't be calling those & wouldn't ever be registered, right?

This discussion reminds me that you should abstract out the get TLS/ check if NULL/ call init_tls into a function named e.g. get_or_create_tls().

Reply:
> So, that gives me an idea that might finesse the issue. In fact I think it already works this way. What if only threads that actually run tasks are registered?

You are correct, it does work this way.

> So an "external" pthread wouldn't be calling those & wouldn't ever be registered, right?

Correct, but by leaving it that way makes any accesses outside of the Chapel tasking layer completely unsafe. For the runtime that is fine since likely only Chapel's managed threads will access that data, but it makes using it for user-data problematic and a bit restrictive. The user could make some adjustments in the management of their data, or better yet just spawn a short-lived Chapel task to obtain data for them (which may throttle potential concurrency and parallelism since it has to wait on a Chapel managed thread...)

Reply:
Is the solution to this problem the enable/disable QSBR calls?

Reply:
Yes.